src/java/org/apache/cassandra/service/DataResurrectionCheck.java

```diff
@@ -0,0 +1,324 @@
+
+        List<Pair<String, String>> violations = new ArrayList<>();
+
+        Set<String> excludedKeyspaces = getExcludedKeyspaces(config);
+        Set<Pair<String, String>> excludedTables = getExcludedTables(config);
+
+        long currentTimeMillis = currentTimeMillis();
+
+        for (String keyspace : getKeyspaces())
+        {
+            if (excludedKeyspaces.contains(keyspace))
+                continue;
+
+            for (TableGCPeriod userTable : getTablesGcPeriods(keyspace))
+            {
+                if (excludedTables.contains(Pair.create(keyspace, userTable.table)))
+                    continue;
+
+                long gcGraceMillis = ((long) userTable.gcPeriod) * 1000;
+                if (heartbeatMillis + gcGraceMillis < currentTimeMillis)
+                    violations.add(Pair.create(keyspace, userTable.table));
+            }
+        }
+
+        if (!violations.isEmpty())
+        {
+            String invalidTables = violations.stream()
+                                             .map(p -> format("%s.%s", p.left, p.right))
+                                             .collect(joining(","));
+
+            String exceptionMessage = format("There are tables for which gcGraceSeconds is older " +
+                                             "then the lastly known time Cassandra node was up based " +
+                                             "on its heartbeat %s with timestamp %s. Cassandra node will not start " +
+                                             "as it would likely introduce data consistency " +
+                                             "issues (zombies etc). Please resolve these issues manually, " +
+                                             "then remove the heartbeat and start the node again. Invalid tables: %s",
+                                             heartbeatFile, heartbeat.lastHeartbeat, invalidTables);
+
+            throw new StartupException(ERR_WRONG_MACHINE_STATE, exceptionMessage);
+        }
+    }
+
+    @Override
+    public void postAction(StartupChecksOptions options)
+    {
+        // Schedule heartbeating after all checks have passed, not as part of the check,
+        // as it might happen that other checks after it might fail, but we would be heartbeating already.
+        if (options.isEnabled(StartupChecks.StartupCheckType.check_data_resurrection))
+        {
+            Map<String, Object> config = options.getConfig(StartupChecks.StartupCheckType.check_data_resurrection);
+            File heartbeatFile = DataResurrectionCheck.getHeartbeatFile(config);
+
+            ScheduledExecutors.scheduledTasks.scheduleAtFixedRate(() -> {
+                Heartbeat heartbeat = new Heartbeat(Instant.ofEpochMilli(Clock.Global.currentTimeMillis()));
+                try
+                {
+                    heartbeatFile.parent().createDirectoriesIfNotExists();
+                    heartbeat.serializeToJsonFile(heartbeatFile);
```

Comment:
can you add a trace log when the heartbeat file is written in case we ever need to debug this?